scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/utils/ReentrantCache.scala

```diff
@@ -0,0 +1,61 @@
+package org.scalafmt.dynamic.utils
+
+import scala.annotation.tailrec
+import scala.concurrent.Await
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.concurrent.duration._
+import scala.util.Try
+
+class ReentrantCache[K, V] {
```

Comment:
Can we add a simple explanation of this data structure? For example "This is like a map except ..."

Reply:
Could we use `Collections.synchronizedMap(...)` and `computeIfAbsent` instead of `ReentrantCache`? To limit synchronization we could have another `HashMap` cache in front of that where we first try with `get()`

Reply:
I agree with @olafurpg; we may we can use Collection.synchronizedMap and `computeIfAbsent` instead.

What we want to do on downloading the ScalafmtReflect (concurrently) is doing the check-then-act operation atomically, and Java's synchronized (or concurrent) collection can seem to play the role.
(ConcurrentHashMap may be better in terms of performance).

Anyway, thank you so much for working on this concurrency issue!

Reply:
I found that `ReentrantCache` is a helpful abstraction of the series of operations (check the result of download and then cache or not)).
Therefore I think it isn't necessarily required to replacing this data structure with built-in data structures (synchronized or concurrent HashMap) as long as the performance doesn't matter.

It would be better to replace this with synchronizedMap or something while keeping the ScalafmtDynamic.scala simple, but I don't have any ideas now...